src/Synthesizer/Modifiers/Envelopes.hs

```diff
@@ -29,25 +31,67 @@ applyEnvelope envelope soundEvent = SoundEvent startTime newEventLength newSampl
                 -- Samples
                 appliedSamples = zipWith (*) envelopeSteps input
                 input = samples samplingRate
+
                 -- ReleaseSamples - release needs the sound after the original samples
-                releaseSamples = zipWith (*) releaseSteps releaseInput
+                releaseSamples =  zipWith (*) releaseSteps releaseInput
                 releaseInput = drop (length envelopeSteps) input
                 releaseSteps = rd
+
                 -- The envelope steps
                 envelopeSteps :: [Double]
                 envelopeSteps = ad ++ dd ++ sd
+
                 sr :: Double
                 sr = fromIntegral samplingRate
-                ad :: [Double] -- [0.0, ..., 1.0]
-                ad = [0.0, (1.0 / (attackLength * sr)) .. 1.0]
-                dd :: [Double] -- [1.0, .., sustainLevel]
-                dd = tail [1.0, 1.0 - ((1.0 - sustainLevel) / (decayLength * sr)) .. sustainLevel]
-                sd :: [Double] --  [sustainLevel]
-                sd = replicate (round (sustainLength * sr)) sustainLevel
-                    where
-                        sustainLength = eventLength - attackLength - decayLength
+
+                -- Calculate the attack steps based on attack length
+                ad :: [Step] -- [0.0, ..., 1.0]
+                ad = getAttackSteps attackLength samplingRate
+
+                -- Calculate the decay steps based on decay length
+                dd :: [Step] -- [1.0, .., sustainLevel]
+                dd = getDecaySteps decayLength sustainLevel samplingRate
+
+                -- Calculate the sustain steps based on sustain level
+                sd :: [Step] --  [sustainLevel]
+                sd = getSustainSteps sustainLevel eventLength attackLength decayLength samplingRate
+
                 -- TODO: the release stepper could start above the sustain level if attack and decay are longer than the eventLength
-                rd :: [Double] -- [sustainLevel, ..., 0.0]
-                rd = tail [sustainLevel, (sustainLevel - step) .. 0.0]
-                    where
-                        step = 1 / (releaseLength * sr) * sustainLevel
+                -- Calculate the release steps based on release length
+                rd :: [Step] -- [sustainLevel, ..., 0.0]
+                rd = getReleaseSteps sustainLevel releaseLength samplingRate
+
+
+
+-- | Calculate the attack steps based on attack length
+getAttackSteps :: AttackLength -> SamplingRate -> [Step]
+getAttackSteps attackLength samplingRate = [0.0, step .. 1.0]
+  where
+    step = 1.0 / (attackLength * sr)
+    sr = fromIntegral samplingRate
```

Comment:
Maybe pass the samplingRate already in a Double instead of converting it in every function